---
title: "ASE simulation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(magrittr)
library(metap)
library(fitdistrplus)
```


```{r}
#' Simulate allele counts for gene expression with an ASE outlier
#' @param n number of samples in simulated cohort  (default = 100)
#' @param mean_het_sites mean number of heterozygous sites in gene (default = 10)
#' @param mu_depth mean read depth of gene, determines mean of read count gamma poisson distribution (default = 100)
#' @param dispersion_depth dispersion paramter of the gamma poisson distribution to simulate gene expression levels (default = 50)
#' @param base_ase expected ASE proportion in non-outlier samples (default = 0.5)
#' @param base_ase_prior_strength determines narrowness and variability of the ASE beta in each sample (higher values ==> tigther dist and less varaince) ( default = 1000)
#' @param outlier_ase the ASE proportion found in the outlier individual
#' @param gene_ase_influence determines narrowness and variability of the gene-level ASE beta when drawing site specific ASE betas (higher values ==> tigther dist and less varaince) ( default = 1000)
ase_outlier_simulation <- function(n = 100, mean_het_sites = 10, mu_depth = 100, dispersion_depth = 50,
                                   base_ase = 0.5, base_ase_prior_strength = 1000, outlier_ase = 0.7,
                                   gene_ase_influence = 1000,
                                  ) {

  ## simulate number of het sites for each sample from poisson
  nhet_sites = rpois(n, mean_het_sites)
  
  ## Simulate read depths from gamma-poisson at each het site
  depths = sapply(1:n, function(x) rnbinom(n = nhet_sites[x], mu = mu_depth, size = dispersion_depth))
  
  ## Simulate gene level ASE beta values from a beta distribution centered at the base_ase param
  ## variance of distriubtion controlled by `base_ase_prior_strength` higher value ==> tighter peak
  sample_betas = rbeta(n = n, base_ase*base_ase_prior_strength, shape2 = (1 - base_ase)*base_ase_prior_strength) 
  
  # randomly select sample to be an outlier
  OUTLIER_IND = sample(1:n, 1)
  sample_betas[OUTLIER_IND] = outlier_ase
  
  # simulate beta values for each site
  ## site specific beta values are drawn from a gene-level beta distribution centered at the gene level ase value
  ## variance of the gene-level beta is determined by gene_ase_influence (higher value ==> tighter dist and smaller variance)
  ## site_level_prior determines the prior distribution of expected site level variance
  
  ## ase beta also randomly flipped with prob 0.5 to account for REF alleles not always being on same haplotype 
  site_betas = lapply(1:n, function(x) abs(rbinom(n = nhet_sites[x], size = 1, prob = .5) -
                                            rbeta(n = nhet_sites[x], shape1 = sample_betas[x]*gene_ase_influence + site_ase_prior,
                                                                     shape2 = (1 - sample_betas[x])*gene_ase_influence + site_ase_prior)))
  
  ## model allele counts at each side from a binomial distribution parameterized by the depths and site_betas
  allele_counts = sapply(1:n, function(x) rbinom(n = nhet_sites[x], prob = site_betas[[x]], size = depths[[x]]))
  ac.df <- NULL
  for (i in 1:length(allele_counts)) {
   ac.df <- rbind(ac.df, data.frame(sample = rep(i, nhet_sites[i]), depth = depths[[i]], REF_count = allele_counts[[i]]))
  }
  ac.df$ALT_count = ac.df$depth - ac.df$REF_count
  ac.df$ase_prop = ac.df$REF_count / ac.df$depth
  ac.df$outlier <- (ac.df$sample == OUTLIER_IND)
  
  ## plot the observed (simulated) ase betas in the outlier compared to the rest of the samples
  print(ggplot(ac.df, aes(ase_prop, fill = outlier)) + geom_histogram() + scale_fill_manual(values = c("grey80", "red")) + theme_classic())
  return (ac.df)
}
  
#' Predict zscores for Allele Specific Expression from a data frame of allele counts
#' @param ac.df allele count data frame, generated by ase_outlier_simulation function
#' @param beta_sample_N number of betas to draw from site level ase betas to form the aggregated gene-level beta estimation (default = 10)
#' @param beta_test_N number of betas to draw when testing a site against the aggregate distribution for outlier status (default = 30)
predict_ase_zscores(ac.df,  beta_sample_N = 10, beta_test_N = 30) {
  ## Create the aggregate meta beta distribution to model gene-level ase from individual site observations
  aggregate_betas <- as.numeric(apply(ac.df, 1, function(row) abs(rbinom(draws,1,.5) - rbeta(beta_sample_N, row[3], row[4]))))
  
  #clunky code to remove beta values that are 0 or 1 (probably a better way)
  # betas of 0.0 and 1.0 cause MLE algorithm to crash
  aggregate_betas[aggregate_betas == 0] <- 1e-3 
  aggregate_betas[aggregate_betas == 1] <- .999
  
  ## fit aggregated betas to a beta distribution
  fit <- fitdist(aggregate_betas, "beta") 
   
  ## plot the estimated aggregate gene-level ase distribution
  print(plot(density(rbeta(10000, fit$estimate[1], fit$estimate[2]))))
  
  # Test samples against the estimate distribution
  sample_p <- NULL
  p.list <- list()
  for (i in 1:n) {
   p_values <- unlist(apply(ac.df[ac.df$sample == i, ], 1, function(row) {
     p = pbeta(rbeta(beta_test_N, row[[3]], row[[4]]), fit$estimate[1], fit$estimate[2])
     mean(pmin(1, 2*pmin(p, 1 - p)))
   }))
   p.list[[i]] = p_values
   # combine p values at each site using fisher's method (metap::sumlog)
   sample_p <- c(sample_p, metap::sumlog(p_values)$p)
  }
  # p-values are right tailed so transform so they look like normal p-values
  ac.df$p_value = 1 - unlist(p.list)
  
  zscores = abs(qnorm(sample_p))
  print(zscores[OUTLIER_IND])
  print(OUTLIER_IND == which.max(zscores))
  
  z.df <- data.frame(sample = 1:length(zscores), z = zscores, logp = -log10(sample_p))
  z.df$outlier =  z.df$sample == OUTLIER_IND
  
  ## plot zscores across sample colored by outlier
  print(ggplot(z.df, aes(sample, logp, color = outlier)) + geom_point() + theme_classic() + scale_color_manual(values = c("grey50", "red")))
}
```

```{r}
ase_outlier_simulation()
```
```{r}
fitdist()
```


